### PR TITLE
Adding Modal component

### LINKED ISSUE #15432: [Ext] Create component: Modal
### Description
Create a reusable UI component for `Modal` according to the insight report

Tasks
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18311
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18161
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18175
- [x] https://github.com/MetaMask/metamask-extension/pull/18979

### References
[Modal Audit FigJam](https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=0%3A1&t=WcWEbEnjEkGuOdXr-1)
[Figma component](https://www.figma.com/file/HKpPKij9V3TpsyMV1TpV7C/DS-Components?node-id=3180%3A13642)
[UI component guidelines](https://www.notion.so/UI-Components-5ccd2bf83eb8441892a0b72c0d8929e1)
[Styling guidelines](https://www.notion.so/Extension-Frontend-Engineering-Guide-b82ddb3e14004b5db3799a9b446294a9#49aa2ef7b7aa40178ff918043df71c34)
Testing guidelines: TBC

### Files needed
- `index.js`
- `index.scss`
- `modal.js`
- `modal.stories.js`
- `modal.test.js`
- `README.mdx`

### Technical details
TBC

### Acceptance criteria
- [ ] Uses color, typography, shadows design tokens 
- [ ] Uses semantic html
- [ ] PropTypes have descriptions
- [ ] Has storybook story with controls
- [ ] Has documentation in MDX
- [ ] Has unit tests and 90% coverage
- [ ] Works in Chrome and Firefox
- [ ] Performance tested: no unnecessary re-renders or other performance concerns

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation

Adding `Modal` component based on [insight report](https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=493%3A233&t=O9QcNdhRvMCxlQIR-1). This PR adds the final components to complete the new `Modal` component as well as stories, documentation and tests. It also introduces a `ModalFocus` component that uses the [react-focus-lock](https://github.com/theKashey/react-focus-lock) library to improve the accessibility of our Modal component for keyboard only users.

* Fixes #15432 
- [Insight report](https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=493%3A233&t=O9QcNdhRvMCxlQIR-1)
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18311
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18161
- [x] Merge https://github.com/MetaMask/metamask-extension/pull/18175
- [x] Add tests

### After

https://user-images.githubusercontent.com/8112138/234746491-46d9f3c5-3c29-4ae5-beba-73c521d422f3.mov

Text coverage for `Modal` and `ModalFocus`

Screenshot 2023-04-27 at 11 56 09 AM

## Manual Testing Steps

- Go to the latest storybook build on this PR
- Search `Modal` in the search bar
- View `stories`, `controls` and `docs`

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: .storybook/preview.js [hunk 1]
```diff
@@ -38,6 +38,9 @@ addParameters({
       ],
     },
   },
+  controls: {
+    expanded: true,
+  },
 });
 
 export const globalTypes = {
```
Comment (georgewrmarshall) [inline .storybook/preview.js:46]: Moved to separate PR https://github.com/MetaMask/metamask-extension/pull/18159

File: ui/components/component-library/badge-wrapper/badge-wrapper.types.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ export enum BadgeWrapperAnchorElementShape {
 
 export interface BadgeWrapperProps
   extends PropTypes.InferProps<typeof Box.propTypes>,
-    React.HTMLAttributes<HTMLDivElement> {
+    React.HTMLAttributes<HTMLElement> {
   /**
    * The element to be wrapped by the BadgeWrapper and for the badge to be positioned on top of
    */
```
Comment (georgewrmarshall) [inline ui/components/component-library/badge-wrapper/badge-wrapper.types.ts:22]: Move to separate PR

File: ui/components/ui/box/box.d.ts [hunk 3]
```diff
@@ -125,6 +130,7 @@ export interface BoxProps {
   style?: React.CSSProperties;
   as?: keyof HTMLElementTagNameMap;
   color?: TextColor | IconColor | string; // TODO: remove string when someone smarter figures out the issue with the color prop
+  ref?: RefObject<HTMLElement>;
 }
 
 declare const Box: React.FC<BoxProps>;
```
Comment (georgewrmarshall) [inline ui/components/ui/box/box.d.ts:136]: Move to another PR

File: ui/components/component-library/modal/README.mdx [hunk 4]
```diff
@@ -0,0 +1,180 @@
+import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
+
+import { Modal } from './modal';
+
+# Modal
+
+The `Modal` is a component that renders a child outside of the normal document flow. It is useful for rendering a child on top of other content.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:7]: Add that it is used together with `ModalOverlay`,`ModalContent` and `ModalHeader`

File: ui/components/component-library/modal/README.mdx [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+### Is Closed On Outside Click
+
+Use the `isClosedOnOutsideClick` prop to control whether the modal should close when the user clicks outside of the modal.
+
+Defaults to `true`.
+
+<Canvas>
+  <Story id="components-componentlibrary-modal--is-closed-on-outside-click" />
+</Canvas>
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:120]: Add code example

File: ui/components/component-library/modal/README.mdx [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+### Is Closed On Escape Key
+
+Use the `isClosedOnEscapeKey` prop to control whether the modal should close when the user presses the escape key.
+
+Defaults to `true`.
+
+<Canvas>
+  <Story id="components-componentlibrary-modal--is-closed-on-escape-key" />
+</Canvas>
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:131]: Add code example

File: ui/components/component-library/modal/README.mdx [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+### Initial Focus Ref
+
+Use the `initialFocusRef` to set the `ref` of the element to receive focus initially. This is useful for input elements that should receive focus when the modal opens.
+
+<Canvas>
+  <Story id="components-componentlibrary-modal--initial-focus-ref" />
+</Canvas>
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:139]: Add code example

File: ui/components/component-library/modal/README.mdx [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+### Final Focus Ref
+
+Use the `finalFocusRef` to set the `ref` of the element to receive focus when the modal closes.
+
+<Canvas>
+  <Story id="components-componentlibrary-modal--final-focus-ref" />
+</Canvas>
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:147]: Add code example

File: ui/components/component-library/modal/README.mdx [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+### Restore Focus
+
+Use the `restoreFocus` prop to restore focus to the element that triggered the `Modal` once it unmounts
+
+<Canvas>
+  <Story id="components-componentlibrary-modal--restore-focus" />
+</Canvas>
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:155]: Add code example

File: ui/components/component-library/modal/README.mdx [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+### Auto Focus
+
+If `true`, the first focusable element within the `children` will auto-focused once `Modal` mounts. Depending on the content of `Modal` this is usually the back or close button in the `ModalHeader`.
+
+Defaults to `true`
+
+<Canvas>
+  <Story id="components-componentlibrary-modal--auto-focus" />
+</Canvas>
+
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:165]: Add code example

File: package.json [hunk 11]
```diff
@@ -333,6 +333,7 @@
     "qrcode.react": "^1.0.1",
     "react": "^16.12.0",
     "react-dom": "^16.12.0",
+    "react-focus-lock": "^2.9.4",
```
Comment (georgewrmarshall) [inline package.json:336]: Using [`react-focus-lock` library](https://github.com/theKashey/react-focus-lock) to lock focus to the `ModalContent` component or to whatever component uses the `modalContentRef` ref to improve the accessibility for keyboard only and screen reader users

File: ui/components/component-library/modal-content/modal-content.tsx [hunk 12]
```diff
@@ -25,6 +25,9 @@ export const ModalContent = ({
       { [`mm-modal-content--size-${size}`]: !width },
       className,
     )}
+    as="section"
+    role="dialog"
+    aria-modal="true"
     backgroundColor={BackgroundColor.backgroundDefault}
     borderRadius={BorderRadius.LG}
     width={width || BLOCK_SIZES.FULL}
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-content/modal-content.tsx:33]: screenreader and semantic html improvements copied from Chakra UI

File: ui/components/component-library/modal-overlay/modal-overlay.tsx [hunk 13]
```diff
@@ -21,6 +21,7 @@ export const ModalOverlay: React.FC<ModalOverlayProps> = ({
     width={BLOCK_SIZES.FULL}
     height={BLOCK_SIZES.FULL}
     onClick={onClick}
+    aria-hidden="true"
     {...props}
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-overlay/modal-overlay.tsx:25]: screenreader improvements copied from Chakra UI

File: ui/components/component-library/modal/README.mdx [hunk 14]
```diff
@@ -0,0 +1,294 @@
+import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
+
+import { Modal } from './modal';
+
+# Modal
+
+The `Modal` focuses the user's attention exclusively on information via a window that is overlaid on primary content. It uses `ModalOverlay`, `ModalContent` and `ModalHeader` to create a modal dialog.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/README.mdx:7]: Change to 
```suggestion
The `Modal` focuses the user's attention exclusively on information via a window that is overlaid on primary content. It should be used with the `ModalOverlay`, `ModalContent` and `ModalHeader` components to create a complete modal.
```

File: ui/components/component-library/modal/modal.scss [hunk 15]
```diff
@@ -0,0 +1,6 @@
+.mm-modal {
+  position: fixed;
+  top: 0;
+  left: 0;
+  // z-index: 1040; // TODO: Check if this is required for Modal to work properly. It aligns with modal z-index in extension
```
Comment (darkwing) [inline ui/components/component-library/modal/modal.scss:5]: Nit: This will need removal.

**Comment by darkwing** [comment]
Should we add a story with large amounts of content so we see how scrolling looks in the modal?

File: ui/components/component-library/modal/README.mdx [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+</Canvas>
+
+```jsx
+import React, { useState, useRef } from 'react';
+
+import { BackgroundColor, BorderRadius, TextColor } from '../../../helpers/constants/design-system';
+
+import Box from '../../ui/box'
+
+import { Modal, ModalOverlay, ModalContent, ModalHeader, Text, Button } from '../../component-library';
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/modal/README.mdx:81]: Good catch thanks! Updated in https://github.com/MetaMask/metamask-extension/pull/18156/commits/627858451d3098088fe9193d1d21824f56ef614a

File: ui/components/component-library/modal/modal.scss (hunk 15 above)
Comment (georgewrmarshall) [reply ui/components/component-library/modal/modal.scss]: Good catch! Will need to address this during the migration to the new component once this has been merged. Removed in https://github.com/MetaMask/metamask-extension/pull/18156/commits/627858451d3098088fe9193d1d21824f56ef614a

**Comment by georgewrmarshall** [comment]
> Should we add a story with large amounts of content so we see how scrolling looks in the modal?

Great point! I think you have identified a BIG BUG with this Modal implementation thank you!!
